### ISSUE TITLE
Estimate variance of correlations between autocorrelated timeseries

### ISSUE THREAD

**Issue by griegner** [nilearn/nilearn#3509]
In reference to #2948, and as discussed during drop-in hours, the temporal dependence in fMRI timeseries can confound measures of functional connectivity. This autocorrelation becomes stronger when applying a masker that averages voxels within brain region, which results in correlating pairs of regions with highly autocorrelated timeseries. This can break the variances-stabilizing property of the Fisher r-to-z transformation, a common intermediate step when thresholding functional connectomes or comparing connectomes across subjects, groups, etc. [Afyouni, Smith, and Nichols (2019)](https://www.sciencedirect.com/science/article/pii/S1053811919303945) proposed a method that accounts for autocorrelation within and between brain regions, and shared their code implementation :) 

I've been experimenting with this code + nilearn and put together an [example notebook](https://github.com/griegner/xDF/blob/gr/Python/01-nilearn-example.ipynb) which reproduces some of the figures in their paper. It's not as user-facing as a nilearn example would be, but I wanted to first see if this methods falls under the scope of the nilearn project. So I'd appreciate any feedback, and in any case, I hope to get more involved with the great work and community here.

**Comment by tsalo** [comment]
It makes sense to me to output variance matrices. xDF has an MIT license, so it should be compatible with the Nilearn BSD-3 license. I don't know the requirements offhand, but we should be able to incorporate the xDF code into Nilearn without too much issue.

**Comment by tsalo** [comment]
>Since there is little use of individual connectivity estimates, it would be interesting to see how it impacts group analyses, e.g. in a classification task, as we do it here to compare correlation vs partial correlation vs tangent

Would this use the corrected z-statistics? It sounds good to me, though I don't know if using test statistics is common for classification tasks.

>What could be clarified before considering inclusion is the benefit

I was also planning to use the variance estimates directly in group-level models, like GAMMs. The specific use-case I was thinking of was when accounting for variable numbers of low-motion volumes after censoring. Generally, people include the number of censored volumes as a confound in the group-level GLM, but after a recent chat with a statistician, it seems like the more appropriate thing to do would be to use the inverse variance estimates as weights.

**Comment by bthirion** [comment]
My understanding is that inverse-variance-weighted estimates and use of censoring information as a a side regressor correspond to 2 different goals: minimum variance estimation vs unconfoundedness. Since I have little experience with the first one, I'm eager to see some empirical evidence of their impact.

**Comment by bthirion** [comment]
OK. I agree that the key for making decisions on these matters is really empirical work.

**Comment by griegner** [comment]
Thank you both for your interest and comments.

> @tsalo I don't know the requirements offhand, but we should be able to incorporate the xDF code into Nilearn without too much issue.

The xDF code only requires NumPy and SciPy, so no added Nilearn dependencies.

> @tsalo The variance could be added as an attribute to a fitted ConnectivityMeasure object, when kind="correlation", maybe? Alternatively, we could refactor xDF (the author is no longer in academia AFAIK) and deploy it to PyPi.  
> @bthirion It seems to be a clear improvement. I would likely implement it a a novel kind of ConnectivityMeasure.

Yes, the ConnectivityMeasure class seems like a good place for it, either adding the variances as an attribute when kind="correlation" is specified or adding a new option under kind="...". Deploying it to PyPi also seems like a good option to gauge interest before adding it to Nilearn.

I had begun to refactor and add docstrings and unit tests as a branch of the [xDF repository](https://github.com/griegner/xDF/tree/gr), so transitioning it to Nilearn or PyPi would be a doable next step for me. 

> It seems to be expensive when used voxel-wise, but we do not encourage people to compute correlations voxel-wise. And I guess that it becomes reasonable when used on a data reduced with a typical atlas.

True, the xDF formula does require the full-lag auto and cross-correlation functions and they are estimated using FFT convolution. This scales efficiently for long timeseries O(N log N), but gets expensive with many pairwise correlations in the voxel-wise case. On my laptop, for timeseries with 1200 timepoints, these calculations take ~1.0 ms per pair. Like you mentioned, the intended application is for data reduced with a typical atlas, in part because this data reduction step increases the autocorrelation of each signal. 

> What could be clarified before considering inclusion is the benefit: Since there is little use of individual connectivity estimates, it would be interesting to see how it impacts group analyses, e.g. in a classification task, as we do it here to compare correlation vs partial correlation vs tangent.

Good point. The xDF paper does focus only on the effects of autocorrelation correction on individual connectivity estimates and graph theory measures, but not on multi-level group models that take these estimates as inputs. For the classification example, here is a quick comparison of the xDF-adjusted z-scores vs the existing methods. Maybe (?) there is a slight improvement in classification accuracy, but this difference between **correlation/naive z-scores** and **xdf-adjusted z-scores** is only because of one missclassified subject. It could be worthwhile to also assess validation and test accuracies on a larger fMRI dataset.

<img src="https://github.com/nilearn/nilearn/assets/54326829/aedc45c8-080b-45b7-bf17-43251a382b3a" width="350">

As an aside, are there some cases in which individual connectivity estimates can be valuable for inferring individual brain networks? I'm curious what your thoughts are -- this paper on [precision functional mapping](https://www.cell.com/neuron/pdf/S0896-6273(17)30613-X.pdf) got me thinking about it.

> @tsalo Would this use the corrected z-statistics? It sounds good to me, though I don't know if using test statistics is common for classification tasks.

My initial thoughts are that applying the xDF r-to-z transformation would benifit any group-level model that assumes its inputs (connectivity estimates) are normally distributed with constant variance. This is because the variance of these correlation estimates depends on the number of timepoints and strength of the correlation, and the transformation will stabilize the variance even under autocorrelation. However, as a recurring theme in this thread, more empirical tests are needed. 

> @tsalo I was also planning to use the variance estimates directly in group-level models, like GAMMs.

Interesting use case. I look forward learning more about it.

**Comment by tsalo** [comment]
>I had begun to refactor and add docstrings and unit tests as a branch of the [xDF repository](https://github.com/griegner/xDF/tree/gr), so transitioning it to Nilearn or PyPi would be a doable next step for me.

Me too! https://github.com/tsalo/xDF/tree/tsalo-edits

I've been chatting with Tom Nichols and there are a couple of issues with applying xDF in common Nilearn use-cases. The big one is that xDF expects a continuous time series, so it won't work well with censored time series. Tom said xDF could be extended to work with these kinds of time series, but it's not available yet.

**Comment by griegner** [comment]
Great! Were you thinking of adding the refactored code to Nilearn or PyPi? Let me know if I can help in any way. 

> The big one is that xDF expects a continuous time series, so it won't work well with censored time series.

I see. Does censoring mean deleting high-motion volumes or including those volumes in a separate "denoising" OLS model as one-hot encoded regressors?

Coincidently, @sjdavenport (who also collaborates with Tom Nichols) and I were thinking of extending the xDF formula to partial correlations. Cool to see that you're also working on new applications -- let's keep in touch on our progress.

**Comment by nicholst** [comment]
Thanks @griegner & others for your efforts on this.  Just to clarify that while @asoroosh has left academia he's still around in a pinch for any clarifications, and I'm certainly able to help.  

As @tsalo mentions, the practical xDF result (the [paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6693558)'s Eq (2), as opposed to the infeasible Eq (1)) assumes stationary time series with an ACF that eventually dies off to zero.  Time series reflecting censoring will strictly violate the stationarity assumption, and creating a version of xDF that accounts for censoring is something I've discussed with @mandymejia, though the practicalities are probably less about the algebra and more about how to retain the efficiency gains of using the FFT.

**Comment by asoroosh** [comment]
It was a delight to see this thread! Amazing work, @griegner, @tsalo, et al.

A couple of notes (FWIW) from me:

- I doubt that xDF is computationally scalable for Voxel-Voxel analysis. However, after we published xDF, we experimented with xDF for voxel-seed analysis ([xDF_SeedFC.m](https://github.com/asoroosh/xDF/blob/master/SeedFC/xDF_seedFC.m)), and it is still computationally manageable on a laptop. The bad news is that it's all in MATLAB, and I never 'translated' it to Python. As far as I can recall, the key differences in the xDF ROI-ROI function are limited to three functions: `xC_fft_sd()`, `SumMat()`, and `ProdMat()`. Also, I would avoid using the xDF function for pairwise analysis. We did our best to use the matrix form as much as possible for speed; hence, it is always faster to avoid pairwise xDF runs.
- AC is a function of ROI size (Appendix E & Fig 2.B of the paper), spatial location (Fig 2.D), smoothing, and sampling (part of a paper in preparation). That being said, I would be cautious (but not disappointed) regarding the cost-benefit of voxel-wise analysis since AC is lower at the voxel level.
- Regarding classification using xDF, I think the impact on accuracy could again be related to item 2: if the majority of AC is explained by these four elements, I wouldn't expect very big accuracy improvement compared to naive z-scores since these four elements remain similar across subjects (but you'd expect significant changes between xDF adjusted and naive at the ROI level). It could turn out to be a simplistic take -- certainly worth a quick experiment.  
- Regarding censoring, if it is done by discarding a time point it violates the assumption of uniform sampling behind FFT, I think the main question would be: if the censoring is performed as @griegner described (i.e. regressed out as a confounder), would it still violates stationarity? I don't think it does.

**Comment by htwangtw** [comment]
Hi! Thanks for the input!

I have spent some time looking at the paper and the code and have come up with some ideas.

I want to see if we can refactor some part of the nilearn `ConnectivityMeasure` class to allow people to pass their own customised estimator. This is the current API:

```python
ConnectivityMeasure(
    cov_estimator=None,  # sklearn estimator class, default: LedoitWolf(store_precision=False)
    kind="covariance",  # The matrix kind, a derivative of covariance
    vectorize=False,
    discard_diagonal=False,The matrix kind
    standardize=True,  # standardise input timeseries
    verbose=0,
)
```

To add a customised estimator it will look like this:

```python

autocorr = AutoCorrelation()  # some customised estimator defined according to scikit-learn guideline

ConnectivityMeasure(
    cov_estimator=autocorr, 
    kind=None,  # we might consider to ignore this argument if a customised estimator is passed
    vectorize=False,
    discard_diagonal=False,The matrix kind
    standardize=True,  # standardise input timeseries
    verbose=0,
)
```

To make this happen we will need to refactor a little bit of the code and add an example based on this paper. I will start a PR to explore this.

cc @Remi-Gau

**Comment by htwangtw** [comment]
Hello! Sorry for the long silence. 

After some further exploration, I conclude that this is mostly a documentation issue. Allowing `None` is pointless, as the default option already passes the matrix as-is. 

Nilearn has an example of using other covariance estimators other than the default already: https://nilearn.github.io/stable/auto_examples/03_connectivity/plot_inverse_covariance_connectome.html#compute-the-sparse-inverse-covariance

I added a little bit more explanation on how `cov_estimator` interacts with `kind` in the API doc. As long as the user passes a covariance estimator in the sklearn API (the example code from @tsalo is very close to what we are looking for). Here's a tutorial on how to create a sklearn estimator on their official docs: https://scikit-learn.org/stable/developers/develop.html

If we want to implement this PR, it has practically become a tutorial on how to write a sklearn estimator. IMO, it is out of scope of nilearn. Here is a potential idea:

@griegner @nicholst can consider refactoring the code based on the scikit-learn tutorial here: https://scikit-learn.org/stable/developers/develop.html and releasing it as a standalone PyPI library. This way, if there are additional ideas for developing more specific estimators to improve functional connectivity, this library can host other estimators written in the same way and compatible with the nilearn API. 

Happy to chat at OHBM if any of you are around ✨

**Comment by bthirion** [comment]
I concur with you. Will be at OHBM too !
